Haiku build is broken in various ways

Tested using `haiku-r1beta3-x86_64-anyboot.iso` (plus the `llvm9_clang` and `CMake` packages installed via HaikuDepot) and the current master branch at commit ff7024d (after my recent CMake change reverted and a couple cosmetic improvements).

### Autoconf 2.69 + GCC 8.3.0: (3 + 7) warnings and 2 errors
```text
$ MATRIX_CMAKE=no MATRIX_CC=gcc ./build_matrix.sh
Haiku shredder 1 hrev55181+66 Jul  6 2022 08:03: x86_64 x86_64 Haiku
OS identification: Haiku-hrev55181
Wed Jul  6 14:00:17 GMT 2022
PREFIX set to '/tmp/libpcap_build_matrix.dgVOcCTs'
===== SETUP 1: CC=gcc CMAKE=no REMOTE=no =====
$ ./build.sh
gcc (2019_05_24) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Compiler identification: gcc-8.3.0
$ ./configure --prefix=/tmp/libpcap_build_matrix.dgVOcCTs --enable-remote=no

[...]

$ make -s
cc1plus: warning: command line option '-Wmissing-prototypes' is valid for C/ObjC but not for C++
cc1plus: warning: command line option '-Wpointer-sign' is valid for C/ObjC but not for C++
cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
pcap-haiku.cpp: In function 'int pcap_read_haiku(pcap_t*, int, pcap_handler, u_char*)':
pcap-haiku.cpp:55:21: warning: unused variable 'handlep' [-Wunused-variable]
  struct pcap_haiku* handlep = (struct pcap_haiku*)handle->priv;
                     ^~~~~~~
pcap-haiku.cpp:50:37: warning: unused parameter 'maxPackets' [-Wunused-parameter]
 pcap_read_haiku(pcap_t* handle, int maxPackets, pcap_handler callback,
                                 ~~~~^~~~~~~~~~
pcap-haiku.cpp: In function 'int pcap_inject_haiku(pcap_t*, const void*, int)':
pcap-haiku.cpp:111:47: warning: unused parameter 'buffer' [-Wunused-parameter]
 pcap_inject_haiku(pcap_t *handle, const void *buffer, int size)
                                   ~~~~~~~~~~~~^~~~~~
pcap-haiku.cpp:111:59: warning: unused parameter 'size' [-Wunused-parameter]
 pcap_inject_haiku(pcap_t *handle, const void *buffer, int size)
                                                       ~~~~^~~~
In file included from pcap-haiku.cpp:12:
pcap-haiku.cpp: In function 'pcap_t* pcap_create_interface(const char*, char*)':
pcap-int.h:494:21: error: types may not be defined in 'sizeof' expressions
      sizeof (struct { pcap_t __common; type __private; }), \
                     ^
pcap-haiku.cpp:252:19: note: in expansion of macro 'PCAP_CREATE_COMMON'
  pcap_t* handle = PCAP_CREATE_COMMON(errorBuffer, struct pcap_haiku);
                   ^~~~~~~~~~~~~~~~~~
In file included from pcap-int.h:37,
                 from pcap-haiku.cpp:12:
pcap-int.h:495:23: error: types may not be defined within __builtin_offsetof
      offsetof (struct { pcap_t __common; type __private; }, __private))
                       ^
pcap-haiku.cpp:252:19: note: in expansion of macro 'PCAP_CREATE_COMMON'
  pcap_t* handle = PCAP_CREATE_COMMON(errorBuffer, struct pcap_haiku);
                   ^~~~~~~~~~~~~~~~~~
pcap-haiku.cpp: In function 'int can_be_bound(const char*)':
pcap-haiku.cpp:268:26: warning: unused parameter 'name' [-Wunused-parameter]
 can_be_bound(const char *name)
              ~~~~~~~~~~~~^~~~
pcap-haiku.cpp: In function 'int get_if_flags(const char*, bpf_u_int32*, char*)':
pcap-haiku.cpp:274:26: warning: unused parameter 'name' [-Wunused-parameter]
 get_if_flags(const char *name, bpf_u_int32 *flags, char *errbuf)
              ~~~~~~~~~~~~^~~~
pcap-haiku.cpp:274:58: warning: unused parameter 'errbuf' [-Wunused-parameter]
 get_if_flags(const char *name, bpf_u_int32 *flags, char *errbuf)
                                                    ~~~~~~^~~~~~
<builtin>: recipe for target 'pcap-haiku.o' failed
make: *** [pcap-haiku.o] Error 1
```

### Autoconf 2.69 + Clang 9.0.1: 5 errors
```text
$ MATRIX_CMAKE=no MATRIX_CC=clang ./build_matrix.sh
Haiku shredder 1 hrev55181+66 Jul  6 2022 08:03: x86_64 x86_64 Haiku
OS identification: Haiku-hrev55181
Wed Jul  6 14:01:12 GMT 2022
PREFIX set to '/tmp/libpcap_build_matrix.v9ft5CHO'
===== SETUP 1: CC=clang CMAKE=no REMOTE=no =====
$ ./build.sh
clang version 9.0.1 
Target: x86_64-unknown-haiku
Thread model: posix
InstalledDir: /bin
Compiler identification: clang-9.0.1
$ ./configure --prefix=/tmp/libpcap_build_matrix.v9ft5CHO --enable-remote=no

[...]

$ make -s
g++: error: unrecognized command line option '-Wcomma'; did you mean '-Wcomment'?
g++: error: unrecognized command line option '-Wdocumentation'
g++: error: unrecognized command line option '-Wmissing-variable-declarations'; did you mean '-Wmissing-declarations'?
g++: error: unrecognized command line option '-Wused-but-marked-unused'
g++: error: unrecognized command line option '-Wshorten-64-to-32'
<builtin>: recipe for target 'pcap-haiku.o' failed
make: *** [pcap-haiku.o] Error 1
```

### CMake 3.23.0 + GCC 8.3.0: (4 + 7) warnings and 2 errors
```text
$ MATRIX_CMAKE=yes MATRIX_CC=gcc ./build_matrix.sh
Haiku shredder 1 hrev55181+66 Jul  6 2022 08:03: x86_64 x86_64 Haiku
OS identification: Haiku-hrev55181
Wed Jul  6 14:03:54 GMT 2022
PREFIX set to '/tmp/libpcap_build_matrix.gB9FeuxR'
===== SETUP 1: CC=gcc CMAKE=yes REMOTE=no =====
$ ./build.sh
gcc (2019_05_24) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Compiler identification: gcc-8.3.0
$ rm -rf CMakeFiles/ CMakeCache.txt
$ mkdir build
$ cd build
$ cmake -DCMAKE_INSTALL_PREFIX=/tmp/libpcap_build_matrix.gB9FeuxR -DENABLE_REMOTE=no ..

[...]

$ make
[  1%] Generating grammar.c, grammar.h
[  3%] Generating scanner.c, scanner.h
[  3%] Built target SerializeTarget
[  5%] Building C object CMakeFiles/pcap.dir/bpf_dump.c.o
[  7%] Building C object CMakeFiles/pcap.dir/bpf_filter.c.o
[  9%] Building C object CMakeFiles/pcap.dir/bpf_image.c.o
[ 10%] Building C object CMakeFiles/pcap.dir/etherent.c.o
[ 12%] Building C object CMakeFiles/pcap.dir/fmtutils.c.o
[ 14%] Building C object CMakeFiles/pcap.dir/gencode.c.o
[ 16%] Building C object CMakeFiles/pcap.dir/nametoaddr.c.o
[ 18%] Building C object CMakeFiles/pcap.dir/optimize.c.o
[ 20%] Building C object CMakeFiles/pcap.dir/pcap-common.c.o
[ 21%] Building C object CMakeFiles/pcap.dir/pcap-options.c.o
[ 23%] Building C object CMakeFiles/pcap.dir/pcap-usb-linux-common.c.o
[ 25%] Building C object CMakeFiles/pcap.dir/pcap.c.o
[ 27%] Building C object CMakeFiles/pcap.dir/savefile.c.o
[ 29%] Building C object CMakeFiles/pcap.dir/sf-pcapng.c.o
[ 30%] Building C object CMakeFiles/pcap.dir/sf-pcap.c.o
[ 32%] Building CXX object CMakeFiles/pcap.dir/pcap-haiku.cpp.o
cc1plus: warning: command line option '-Wmissing-prototypes' is valid for C/ObjC but not for C++
cc1plus: warning: command line option '-Wpointer-sign' is valid for C/ObjC but not for C++
cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
cc1plus: warning: command line option '-std=gnu99' is valid for C/ObjC but not for C++
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'int pcap_read_haiku(pcap_t*, int, pcap_handler, u_char*)':
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:55:21: warning: unused variable 'handlep' [-Wunused-variable]
  struct pcap_haiku* handlep = (struct pcap_haiku*)handle->priv;
                     ^~~~~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:50:37: warning: unused parameter 'maxPackets' [-Wunused-parameter]
 pcap_read_haiku(pcap_t* handle, int maxPackets, pcap_handler callback,
                                 ~~~~^~~~~~~~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'int pcap_inject_haiku(pcap_t*, const void*, int)':
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:111:47: warning: unused parameter 'buffer' [-Wunused-parameter]
 pcap_inject_haiku(pcap_t *handle, const void *buffer, int size)
                                   ~~~~~~~~~~~~^~~~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:111:59: warning: unused parameter 'size' [-Wunused-parameter]
 pcap_inject_haiku(pcap_t *handle, const void *buffer, int size)
                                                       ~~~~^~~~
In file included from /boot/home/gitrepos/libpcap/pcap-haiku.cpp:12:
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'pcap_t* pcap_create_interface(const char*, char*)':
/boot/home/gitrepos/libpcap/pcap-int.h:494:21: error: types may not be defined in 'sizeof' expressions
      sizeof (struct { pcap_t __common; type __private; }), \
                     ^
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:252:19: note: in expansion of macro 'PCAP_CREATE_COMMON'
  pcap_t* handle = PCAP_CREATE_COMMON(errorBuffer, struct pcap_haiku);
                   ^~~~~~~~~~~~~~~~~~
In file included from /boot/home/gitrepos/libpcap/pcap-int.h:37,
                 from /boot/home/gitrepos/libpcap/pcap-haiku.cpp:12:
/boot/home/gitrepos/libpcap/pcap-int.h:495:23: error: types may not be defined within __builtin_offsetof
      offsetof (struct { pcap_t __common; type __private; }, __private))
                       ^
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:252:19: note: in expansion of macro 'PCAP_CREATE_COMMON'
  pcap_t* handle = PCAP_CREATE_COMMON(errorBuffer, struct pcap_haiku);
                   ^~~~~~~~~~~~~~~~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'int can_be_bound(const char*)':
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:268:26: warning: unused parameter 'name' [-Wunused-parameter]
 can_be_bound(const char *name)
              ~~~~~~~~~~~~^~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'int get_if_flags(const char*, bpf_u_int32*, char*)':
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:274:26: warning: unused parameter 'name' [-Wunused-parameter]
 get_if_flags(const char *name, bpf_u_int32 *flags, char *errbuf)
              ~~~~~~~~~~~~^~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:274:58: warning: unused parameter 'errbuf' [-Wunused-parameter]
 get_if_flags(const char *name, bpf_u_int32 *flags, char *errbuf)
                                                    ~~~~~~^~~~~~
CMakeFiles/pcap.dir/build.make:299: recipe for target 'CMakeFiles/pcap.dir/pcap-haiku.cpp.o' failed
make[2]: *** [CMakeFiles/pcap.dir/pcap-haiku.cpp.o] Error 1
CMakeFiles/Makefile2:159: recipe for target 'CMakeFiles/pcap.dir/all' failed
make[1]: *** [CMakeFiles/pcap.dir/all] Error 2
Makefile:135: recipe for target 'all' failed
make: *** [all] Error 2
```

### CMake 3.23.0 + Clang 9.0.1: (4 + 7) warnings and 2 errors
```text
$ MATRIX_CMAKE=yes MATRIX_CC=clang ./build_matrix.sh
Haiku shredder 1 hrev55181+66 Jul  6 2022 08:03: x86_64 x86_64 Haiku
OS identification: Haiku-hrev55181
Wed Jul  6 14:04:59 GMT 2022
PREFIX set to '/tmp/libpcap_build_matrix.C3wftnfj'
===== SETUP 1: CC=clang CMAKE=yes REMOTE=no =====
$ ./build.sh
clang version 9.0.1 
Target: x86_64-unknown-haiku
Thread model: posix
InstalledDir: /bin
Compiler identification: clang-9.0.1
$ rm -rf CMakeFiles/ CMakeCache.txt
$ cd build
$ cmake -DCMAKE_INSTALL_PREFIX=/tmp/libpcap_build_matrix.C3wftnfj -DENABLE_REMOTE=no ..

[...]

$ make
[  1%] Generating grammar.c, grammar.h
[  3%] Generating scanner.c, scanner.h
[  3%] Built target SerializeTarget
[  5%] Building C object CMakeFiles/pcap.dir/bpf_dump.c.o
[  7%] Building C object CMakeFiles/pcap.dir/bpf_filter.c.o
[  9%] Building C object CMakeFiles/pcap.dir/bpf_image.c.o
[ 10%] Building C object CMakeFiles/pcap.dir/etherent.c.o
[ 12%] Building C object CMakeFiles/pcap.dir/fmtutils.c.o
[ 14%] Building C object CMakeFiles/pcap.dir/gencode.c.o
[ 16%] Building C object CMakeFiles/pcap.dir/nametoaddr.c.o
[ 18%] Building C object CMakeFiles/pcap.dir/optimize.c.o
[ 20%] Building C object CMakeFiles/pcap.dir/pcap-common.c.o
[ 21%] Building C object CMakeFiles/pcap.dir/pcap-options.c.o
[ 23%] Building C object CMakeFiles/pcap.dir/pcap-usb-linux-common.c.o
[ 25%] Building C object CMakeFiles/pcap.dir/pcap.c.o
[ 27%] Building C object CMakeFiles/pcap.dir/savefile.c.o
[ 29%] Building C object CMakeFiles/pcap.dir/sf-pcapng.c.o
[ 30%] Building C object CMakeFiles/pcap.dir/sf-pcap.c.o
[ 32%] Building CXX object CMakeFiles/pcap.dir/pcap-haiku.cpp.o
cc1plus: warning: command line option '-Wmissing-prototypes' is valid for C/ObjC but not for C++
cc1plus: warning: command line option '-Wpointer-sign' is valid for C/ObjC but not for C++
cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
cc1plus: warning: command line option '-std=gnu99' is valid for C/ObjC but not for C++
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'int pcap_read_haiku(pcap_t*, int, pcap_handler, u_char*)':
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:55:21: warning: unused variable 'handlep' [-Wunused-variable]
  struct pcap_haiku* handlep = (struct pcap_haiku*)handle->priv;
                     ^~~~~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:50:37: warning: unused parameter 'maxPackets' [-Wunused-parameter]
 pcap_read_haiku(pcap_t* handle, int maxPackets, pcap_handler callback,
                                 ~~~~^~~~~~~~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'int pcap_inject_haiku(pcap_t*, const void*, int)':
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:111:47: warning: unused parameter 'buffer' [-Wunused-parameter]
 pcap_inject_haiku(pcap_t *handle, const void *buffer, int size)
                                   ~~~~~~~~~~~~^~~~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:111:59: warning: unused parameter 'size' [-Wunused-parameter]
 pcap_inject_haiku(pcap_t *handle, const void *buffer, int size)
                                                       ~~~~^~~~
In file included from /boot/home/gitrepos/libpcap/pcap-haiku.cpp:12:
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'pcap_t* pcap_create_interface(const char*, char*)':
/boot/home/gitrepos/libpcap/pcap-int.h:494:21: error: types may not be defined in 'sizeof' expressions
      sizeof (struct { pcap_t __common; type __private; }), \
                     ^
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:252:19: note: in expansion of macro 'PCAP_CREATE_COMMON'
  pcap_t* handle = PCAP_CREATE_COMMON(errorBuffer, struct pcap_haiku);
                   ^~~~~~~~~~~~~~~~~~
In file included from /boot/home/gitrepos/libpcap/pcap-int.h:37,
                 from /boot/home/gitrepos/libpcap/pcap-haiku.cpp:12:
/boot/home/gitrepos/libpcap/pcap-int.h:495:23: error: types may not be defined within __builtin_offsetof
      offsetof (struct { pcap_t __common; type __private; }, __private))
                       ^
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:252:19: note: in expansion of macro 'PCAP_CREATE_COMMON'
  pcap_t* handle = PCAP_CREATE_COMMON(errorBuffer, struct pcap_haiku);
                   ^~~~~~~~~~~~~~~~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'int can_be_bound(const char*)':
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:268:26: warning: unused parameter 'name' [-Wunused-parameter]
 can_be_bound(const char *name)
              ~~~~~~~~~~~~^~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp: In function 'int get_if_flags(const char*, bpf_u_int32*, char*)':
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:274:26: warning: unused parameter 'name' [-Wunused-parameter]
 get_if_flags(const char *name, bpf_u_int32 *flags, char *errbuf)
              ~~~~~~~~~~~~^~~~
/boot/home/gitrepos/libpcap/pcap-haiku.cpp:274:58: warning: unused parameter 'errbuf' [-Wunused-parameter]
 get_if_flags(const char *name, bpf_u_int32 *flags, char *errbuf)
                                                    ~~~~~~^~~~~~
CMakeFiles/pcap.dir/build.make:299: recipe for target 'CMakeFiles/pcap.dir/pcap-haiku.cpp.o' failed
make[2]: *** [CMakeFiles/pcap.dir/pcap-haiku.cpp.o] Error 1
CMakeFiles/Makefile2:159: recipe for target 'CMakeFiles/pcap.dir/all' failed
make[1]: *** [CMakeFiles/pcap.dir/all] Error 2
Makefile:135: recipe for target 'all' failed
make: *** [all] Error 2
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -580,6 +580,15 @@ else(WIN32)
             set(PCAP_LINK_LIBRARIES str ${PCAP_LINK_LIBRARIES})
         endif(LIBSTR_HAS_PUTMSG)
     endif(NOT STDLIBS_HAVE_PUTMSG)
+
+    # Haiku has getpass in libbsd
+    check_function_exists(getpass STDLIBS_HAVE_GETPASS)
+    if(NOT STDLIBS_HAVE_GETPASS)
+        check_library_exists(bsd getpass "" LIBBSD_HAS_GETPASS)
+        if(LIBBSD_HAS_GETPASS)
+            set(PCAP_LINK_LIBRARIES bsd ${PCAP_LINK_LIBRARIES})
+        endif(LIBBSD_HAS_GETPASS)
+    endif(NOT STDLIBS_HAVE_GETPASS)
 endif(WIN32)
 
 #

--- a/configure.ac
+++ b/configure.ac
@@ -70,6 +70,10 @@ haiku*)
 	# Haiku needs _BSD_SOURCE for the _IO* macros because it doesn't use them.
 	#
 	CFLAGS="$CFLAGS -D_BSD_SOURCE"
+	#
+	# Haiku has getpass in libbsd.
+	#
+	AC_CHECK_LIB(bsd, getpass)
 	;;
 esac
 

--- a/pcap-haiku.cpp
+++ b/pcap-haiku.cpp
@@ -249,7 +249,11 @@ pcap_create_interface(const char *device, char *errorBuffer)
 		return NULL;
 	}
 
-	pcap_t* handle = PCAP_CREATE_COMMON(errorBuffer, struct pcap_haiku);
+	struct wrapper_struct { pcap_t __common; struct pcap_haiku __private; };
+	pcap_t* handle = pcap_create_common(errorBuffer,
+		sizeof (struct wrapper_struct),
+		offsetof (struct wrapper_struct, __private));
+
 	if (handle == NULL) {
 		snprintf(errorBuffer, PCAP_ERRBUF_SIZE, "malloc: %s", strerror(errno));
 		close(socket);

--- a/rpcap-protocol.h
+++ b/rpcap-protocol.h
@@ -132,10 +132,12 @@
  * XXX - use the C99 types?  Microsoft's newer versions of Visual Studio
  * support them.
  */
+#ifndef __HAIKU__
 typedef unsigned char uint8;	/* 8-bit unsigned integer */
 typedef unsigned short uint16;	/* 16-bit unsigned integer */
 typedef unsigned int uint32;	/* 32-bit unsigned integer */
 typedef int int32;		/* 32-bit signed integer */
+#endif
 
 /* Common header for all the RPCAP messages */
 struct rpcap_header

